### PR TITLE
[random] add support for enums

### PR DESCRIPTION
**Pull request by Wh1teDuke**

### CODE REVIEW

File: lib/pure/random.nim [hunk 1]
```diff
@@ -162,9 +165,13 @@ proc rand*[T](r: var Rand; a: openArray[T]): T {.deprecated.} =
   ## **Deprecated since v0.20.0:** use ``sample`` instead.
   result = a[rand(r, a.low..a.high)]
 
-proc rand*[T: SomeInteger](t: typedesc[T]): T =
-  ## Returns a random integer in the range `low(T)..high(T)`.
-  result = cast[T](state.next)
+proc rand*[T: SomeInteger | enum](r: var Rand; t: typedesc[T]): T =
+  ## Returns a random element in the range `low(T)..high(T)`.
+  result = cast[T](r.next)
```
Comment (GULPF) [inline lib/pure/random.nim]: This isn't correct for enums. `cast` will give a random value with the correct bit size, but not necessarily within the valid range for the enum.

File: lib/pure/random.nim [hunk 2]
```diff
@@ -162,9 +165,17 @@ proc rand*[T](r: var Rand; a: openArray[T]): T {.deprecated.} =
...
-proc rand*[T: SomeInteger](t: typedesc[T]): T =
-  ## Returns a random integer in the range `low(T)..high(T)`.
-  result = cast[T](state.next)
+proc rand*[T: SomeInteger](r: var Rand; t: typedesc[T]): T =
+  ## Returns a random element in the range `low(T)..high(T)`.
+  result = cast[T](r.next)
+
+proc rand*[T: enum](r: var Rand; t: typedesc[T]): T =
+  ## Returns a random element in the range `low(T)..high(T)`.
+  result = T(rand(r, ord(high(T))))
```
Comment (GULPF) [inline lib/pure/random.nim]: This is still incorrect for enums that doesn't start at zero. `rand(r, low(T)..high(T))` will work except for enums with holes, but that might not matter.

Comment (mratsim) [reply lib/pure/random.nim]: Maybe we should have a typetraits that test if enum has hole and statically fail for that case?

Comment (Araq) [reply lib/pure/random.nim]: Enums with holes are bad, we need to remove them from the language.

**Comment by Araq** [comment]
It's hard to get right, better not support it. One can always use ``E(randNumer)`` to convert back to the enum.

**Pull request closed without merging**